migrations/versions/0476_rename_contacted_users.py

```diff
@@ -0,0 +1,59 @@
+"""
+Create Date: 2024-11-18
+"""
+
+from alembic import op
+import sqlalchemy as sa
+from sqlalchemy.dialects import postgresql
+
+revision = "0476_rename_contacted_users"
+down_revision = "0475_ntfcn_hist_xstats_mcv"
+
+
+def upgrade():
+    # 1. Create service_join_request_contacted_users
+    op.create_table(
+        "service_join_request_contacted_users",
+        sa.Column(
+            "service_join_request_id",
+            postgresql.UUID(as_uuid=True),
+            sa.ForeignKey("service_join_requests.id"),
+            primary_key=True,
+        ),
+        sa.Column("user_id", postgresql.UUID(as_uuid=True), sa.ForeignKey("users.id"), primary_key=True),
+    )
+
+    # 2. Copy data from the old table to the new
+    op.execute(
+        """
+        INSERT INTO service_join_request_contacted_users (service_join_request_id, user_id)
+        SELECT service_join_request_id, user_id FROM contacted_users;
+        """
+    )
+
+    # 3. Drop the old table
+    op.drop_table("contacted_users")
```

Comment:
i think these two lines (and the corresponding migrations) should be in a separate migration.

otherwise, while this PR is deploying, we run migrations first, and then in the ~10 minutes or so it takes to roll instances, Notify will break and serve error messages as it will try and read from `contacted_users` but the table no longer exists

Reply:
this means that the data will be temporarily empty between the two migrations - but i think that's acceptable to do while this feature isn't used. but i'd rather ensure we never return 5xx errors

Reply:
Hi Leo, actually after talking with Katie and Saurabh the following has been suggested:

https://github.com/alphagov/notifications-api/pull/4265/commits/3adc11f21ea0ad045fa72fe4e8f0654f8f560c2b

Renaming the table instead of the dropping approach - this is the resource @spatel033 gave me: https://petegraham.co.uk/rename-postgres-table-with-alembic/

Let me know what you think

Reply:
Concourse lint-migration is failing due to renaming the table. There is an extra step is required to rename table with minor lock: https://squawkhq.com/docs/renaming-table

Reply:
i guess the advantage of the view method is that it handles data transfer and ensuring reads/writes are uninterrupted. but it doesn't solve the issues from that blog where we'd want to ensure we rename indexes/sequences/etc

i'm easy on which one we choose - but either way i think two PRs feels inevitable

Reply:
wait, you can insert to a view if its only got one from table 🤯 

every day's a school day!